### PR TITLE
feat(reactions): New button for web.

### PR DESCRIPTION
**Pull request by hristoterezov**

### CODE REVIEW

File: css/_reactions-menu.scss [hunk 1]
```diff
@@ -1,5 +1,31 @@
 @use 'sass:math';
 
+#overflow-context-menu {
```
Comment (saghul) [inline css/_reactions-menu.scss:3]: Can we move this to JSS? @robertpin has been moving SCSS code to JSS for a while, it would be nice not to add more.

File: react/features/base/config/configType.ts [hunk 2]
```diff
@@ -462,6 +462,7 @@ export interface IConfig {
         hideExtraJoinButtons?: Array<string>;
     };
     prejoinPageEnabled?: boolean;
+    reactionsButtonEnabled?: boolean;
```
Comment (saghul) [inline react/features/base/config/configType.ts:465]: Can we use `toolbarButtons` instead? That is how all buttons are currently configured, why does this one need a dedicated config option?

File: react/features/base/toolbox/components/web/ToolboxButtonWithIconPopup.tsx [hunk 3]
```diff
@@ -33,17 +33,17 @@ interface IProps {
     /**
      * Icon of the button.
      */
-    icon: Function;
+    icon?: Function;
```
Comment (saghul) [inline react/features/base/toolbox/components/web/ToolboxButtonWithIconPopup.tsx:36]: This component is called ToolboxButtonWithIconPopup, making the icon optional feels wrong.

File: react/features/base/popover/components/Popover.web.tsx [hunk 4]
```diff
@@ -304,7 +304,8 @@ class Popover extends Component<IProps, IState> {
             && !this.props.overflowDrawer
             && this._contextMenuRef
             && this._contextMenuRef.contains
-            && !this._contextMenuRef.contains(event.target as Node)) {
+            && !this._contextMenuRef.contains(event.target as Node)
+            && !this._containerRef?.current?.contains(event.target as Node)) {
```
Comment (saghul) [inline react/features/base/popover/components/Popover.web.tsx:308]: Why is this necessary?

File: react/features/gifs/actions.ts [hunk 5]
```diff
@@ -86,3 +87,16 @@ export function setGifMenuVisibility(visible: boolean) {
         visible
     };
 }
+
+/**
+ * Set visibility of the GIF drawer.
```
Comment (saghul) [inline react/features/gifs/actions.ts:92]: Pl adjust the comment.

File: react/features/gifs/actions.ts [hunk 6]
```diff
@@ -86,3 +87,16 @@ export function setGifMenuVisibility(visible: boolean) {
...
+ *
+ * @param {boolean} visible - Whether or not it should be visible.
+ * @returns {Object}
+ */
+export function setGifOverflowMenuVisibility(visible: boolean) {
```
Comment (saghul) [inline react/features/gifs/actions.ts:97]: Why are we calling this is an overflow menu? Is it not a popover?

File: react/features/gifs/components/web/GifsMenuButton.tsx [hunk 7]
```diff
@@ -4,10 +4,14 @@ import { useDispatch, useSelector } from 'react-redux';
 
 import ReactionButton from '../../../reactions/components/web/ReactionButton';
 import { showOverflowDrawer } from '../../../toolbox/functions.web';
-import { setGifDrawerVisibility, setGifMenuVisibility } from '../../actions';
+import { setGifDrawerVisibility, setGifMenuVisibility, setGifOverflowMenuVisibility } from '../../actions';
 import { isGifsMenuOpen } from '../../functions.web';
 
-const GifsMenuButton = () => {
+interface IProps {
+    overflowMenu?: boolean;
```
Comment (saghul) [inline react/features/gifs/components/web/GifsMenuButton.tsx:11]: Can't we use useState instead? Feels weird to get the overflow menu and overflow drawer options differently.

File: react/features/gifs/functions.web.ts [hunk 8]
```diff
@@ -11,7 +11,12 @@ export * from './function.any';
  */
 export function isGifsMenuOpen(state: IReduxState) {
     const overflowDrawer = showOverflowDrawer(state);
-    const { drawerVisible, menuOpen } = state['features/gifs'];
+    const { overflowMenuVisible } = state['features/toolbox'];
```
Comment (saghul) [inline react/features/gifs/functions.web.ts:14]: Why do we need to check if the overflow menu is visible? Isn't it enough to check if the gif menu is open?

File: react/features/gifs/reducer.ts [hunk 9]
```diff
@@ -23,6 +25,7 @@ export interface IGifsState {
     drawerVisible: boolean;
     gifList: Map<string, IGif>;
     menuOpen: boolean;
+    overflowVisible: boolean;
```
Comment (saghul) [inline react/features/gifs/reducer.ts:28]: Wouldn't it make more sense to turn these 2 variables into a `mode` enum? The mode would be either `null`, `drawer` or `overflow`.

File: react/features/reactions/components/web/ReactionsMenu.tsx [hunk 10]
```diff
@@ -56,13 +50,43 @@ interface IProps {
...
+const useStyles = makeStyles<IProps>()((theme, props: IProps) => {
+    const { overflowMenu, overflowDrawer, showRaisedHand, _isGifMenuVisible } = props;
+    let reactionsMenuHeight = REACTIONS_MENU_HEIGHT;
+
+    if (overflowMenu) {
+        if (!showRaisedHand) {
+            reactionsMenuHeight -= RAISE_HAND_ROW_HEIGHT;
+        }
+        if (!overflowDrawer) {
+            reactionsMenuHeight -= 38;
```
Comment (saghul) [inline react/features/reactions/components/web/ReactionsMenu.tsx:83]: Create a constant pl

File: react/features/reactions/components/web/ReactionsMenu.tsx [hunk 11]
```diff
@@ -56,13 +50,43 @@ interface IProps {
...
+        }
+        if (!overflowDrawer && _isGifMenuVisible) {
+            reactionsMenuHeight += 200;
```
Comment (saghul) [inline react/features/reactions/components/web/ReactionsMenu.tsx:86]: Ditto

File: react/features/toolbox/components/web/OverflowMenuButton.tsx [hunk 12]
```diff
@@ -29,46 +34,83 @@ interface IProps {
     /**
      * A child React Element to display within {@code Popover}.
      */
-    children: ReactNode;
+    children: Object[];
```
Comment (saghul) [inline react/features/toolbox/components/web/OverflowMenuButton.tsx:37]: Why do you need to change this type?

File: react/features/toolbox/components/web/OverflowMenuButton.tsx [hunk 13]
```diff
@@ -88,50 +130,120 @@ const OverflowMenuButton = ({
...
+    let reactionsMenuHeight = 0;
+
+    if (showReactionsMenu) {
+        reactionsMenuHeight = REACTIONS_MENU_HEIGHT;
+
+        if (!showRaiseHandInReactionsMenu) {
+            reactionsMenuHeight -= RAISE_HAND_ROW_HEIGHT;
+        }
+        if (!overflowDrawer) {
+            reactionsMenuHeight -= 38;
```
Comment (saghul) [inline react/features/toolbox/components/web/OverflowMenuButton.tsx:144]: This needs a constant.

File: react/features/toolbox/components/web/OverflowMenuButton.tsx [hunk 14]
```diff
@@ -88,50 +130,120 @@ const OverflowMenuButton = ({
...
+        }
+        if (!overflowDrawer && isGiphyVisible) {
+            reactionsMenuHeight += 200;
```
Comment (saghul) [inline react/features/toolbox/components/web/OverflowMenuButton.tsx:147]: Ditto.

File: react/features/toolbox/components/web/Toolbox.tsx [hunk 15]
```diff
@@ -1011,6 +1030,11 @@ class Toolbox extends Component<IProps> {
             !_jwtDisabledButons.includes(key)
             && (isToolbarButtonEnabled(key, _toolbarButtons) || isToolbarButtonEnabled(alias, _toolbarButtons))
         );
+        const filteredKeys = filtered.map(button => button.key);
```
Comment (saghul) [inline react/features/toolbox/components/web/Toolbox.tsx:1033]: Why is this logic necessary?

File: react/features/toolbox/components/web/Toolbox.tsx [hunk 16]
```diff
@@ -1427,44 +1458,34 @@ class Toolbox extends Component<IProps> {
...
-
-                                            if (group === val.group) {
-                                                prev.push(val);
-                                            } else {
-                                                acc.push([ val ]);
-                                            }
+                                showRaiseHandInReactionsMenu = { showRaiseHandInReactionsMenu }
+                                showReactionsMenu = { showReactionsInOverflowMenu }>
+                                {overflowMenuButtons.reduce((acc, val) => {
+                                    if (val.key === 'reactions' && showReactionsInOverflowMenu) {
```
Comment (saghul) [inline react/features/toolbox/components/web/Toolbox.tsx:1466]: Why do these need special treatment now?

**Review by saghul — COMMENTED** [review]
I took a look and left some questions! Have you deployed this somewhere to test? I'd also like @robertpin to take a look here.

**Comment by robertpin** [comment]
@hristoterezov can you add some screenshots and some details? I'm kinda out of the loop so I don't know what/ why we're doing this.

**Comment by hristoterezov** [comment]
> @hristoterezov can you add some screenshots and some details? I'm kinda out of the loop so I don't know what/ why we're doing this.

@robertpin It introduces a new separate reactions button and covers all the corner cases around it (for example implement reactions in the non-drawer overflow menu mode). The motivation for the change was that some users are clicking the raise hand button by mistake when they try to open the reactions menu.

Screenshots:
image
image

**Comment by hristoterezov** [comment]
> I took a look and left some questions! Have you deployed this somewhere to test? I'd also like @robertpin to take a look here.

@saghul It is deployed on hristo.jitsi.net

File: css/_reactions-menu.scss (hunk 1 above)
Comment (hristoterezov) [reply css/_reactions-menu.scss:3]: will give it a try

File: react/features/base/config/configType.ts (hunk 2 above)
Comment (hristoterezov) [reply react/features/base/config/configType.ts:465]: I was considering this since you initially proposed it and to be honest I really like the idea. The downside is that on mobile we don't use them and that's why I decided to go with the config option. Maybe we can start using the `config.toolbarButtons` on mobile and initially use it only for the reactions button. WDYT?

File: react/features/gifs/reducer.ts (hunk 9 above)
Comment (hristoterezov) [reply react/features/gifs/reducer.ts:28]: Good idea! Will give it a try!

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:37]: Because now we are not constructing the JSX outside of the OverflowMenuButton but inside it.

File: react/features/toolbox/components/web/Toolbox.tsx (hunk 15 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/Toolbox.tsx:1033]: In order to handle the scenario where a button defined in `order` turns out to be disabled. In this case we need to calculate how much buttons we need to display on the main toolbar. If we go with the previous implementation some random buttons from the overflow menu are actually displayed on the main toolbar.

File: react/features/base/config/configType.ts (hunk 2 above)
Comment (saghul) [reply react/features/base/config/configType.ts:465]: We can try that! I thought this PR didn't tackle mobile yet though.

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (saghul) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:37]: Not sure I understand. What type are the children now? I haven't seen that change.

File: react/features/toolbox/components/web/Toolbox.tsx (hunk 15 above)
Comment (saghul) [reply react/features/toolbox/components/web/Toolbox.tsx:1033]: I see. Isn't that a separate change from this though?

File: react/features/base/config/configType.ts (hunk 2 above)
Comment (hristoterezov) [reply react/features/base/config/configType.ts:465]: It doesn't but there will be a follow up one for mobile.

Will do the change!

File: react/features/toolbox/components/web/Toolbox.tsx (hunk 16 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/Toolbox.tsx:1466]: If we need to show reactions in the overflow menu we filter them out of the array with buttons just for convenience because we have separate property anyway and we display them differently than the other buttons. In theory we can probably move this logic into OverflowMenuButton but we were doing this for the raise hand already anyway and also we do the grouping here and iterate over the array so I thought it is not a bad idea to also filter them here.

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:37]: They are array with objects. Those are the grouped objects which are representing the buttons from where we transform them in JSX. Check Toolbox.tsx. Before we were building the whole JSX for the content of the overflow menu. I've moved the JSX building here.

File: react/features/toolbox/components/web/Toolbox.tsx (hunk 15 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/Toolbox.tsx:1033]: It could be. That was existing problem even before. I will move it in a separate commit.

File: react/features/gifs/actions.ts (hunk 6 above)
Comment (hristoterezov) [reply react/features/gifs/actions.ts:97]: It is but the component was called OverflowMenuButton. I don't have any preference here on the naming. If you want me I can rename it somehow.

File: react/features/base/toolbox/components/web/ToolboxButtonWithIconPopup.tsx (hunk 3 above)
Comment (hristoterezov) [reply react/features/base/toolbox/components/web/ToolboxButtonWithIconPopup.tsx:36]: Yep. Will think how to change it. Will probably create a ToolboxButtonWithPopup and reuse it here.

File: react/features/gifs/components/web/GifsMenuButton.tsx (hunk 7 above)
Comment (hristoterezov) [reply react/features/gifs/components/web/GifsMenuButton.tsx:11]: Good catch!

I believe what happened here is during the time I was resolving some conflicts something changed and I haven't resolved them correctly because we are still passing the overflow drawer prop directly to the component but in the same time we don't use it and use the selector.

File: react/features/base/popover/components/Popover.web.tsx (hunk 4 above)
Comment (robertpin) [reply react/features/base/popover/components/Popover.web.tsx:308]: Did you set the Popover `trigger` prop to `click`?

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (robertpin) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:37]: In Toolbox, the OverflowMenuButton currently only has one child: ContextMenu. And it's definitely JSX, not an Object

File: react/features/base/popover/components/Popover.web.tsx (hunk 4 above)
Comment (hristoterezov) [reply react/features/base/popover/components/Popover.web.tsx:308]: yes

File: react/features/gifs/functions.web.ts (hunk 8 above)
Comment (hristoterezov) [reply react/features/gifs/functions.web.ts:14]: I was actually doing the same as we do for the drawer. 

On the second look I think we don't need checks at all. I changed it to check only for `menuOpen` even for the drawer case. The only side effect is if you change layouts and we go from reactions displayed in drawer to overflow menu to the main toolbar the state of the gif menu will be preserved which IMHO is not bad and this is kind of a corner case.

Good catch! This simplifies the logic a lot.

File: react/features/gifs/reducer.ts (hunk 9 above)
Comment (hristoterezov) [reply react/features/gifs/reducer.ts:28]: This is related to the previous reply. Now `overflowVisible` and `drawerVisible` are gone because we don't need them anymore!

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:43]: @robertpin See this change. https://github.com/jitsi/jitsi-meet/pull/13211/files#diff-3e1fb4316fab0334bc37d37a0f046777e7ce3fd8c8e4321cc8951edbffe2b972L1432 . IMHO this is definitely an array of Objects.

Am I missing something?

Comment (robertpin) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:43]: Oh, I didn't see you changed it. Hmm, seems really weird to pass some Objects as children. Why do we do this?

File: react/features/gifs/components/web/GifsMenu.tsx [hunk 17]
```diff
@@ -77,19 +83,25 @@ const useStyles = makeStyles()(theme => {
     };
 });
 
+interface IProps {
+    columns?: number;
+    parent: IReactionsMenuParent;
```
Comment (robertpin) [inline react/features/gifs/components/web/GifsMenu.tsx:88]: This feels weird. Normally in React the child component shouldn't know about the parent component. Can't we just pass an `inDrawer` prop or something?

File: react/features/reactions/components/web/ReactionsMenuButton.tsx [hunk 18]
```diff
@@ -60,13 +69,28 @@ interface IProps extends WithTranslation {
     reactionsQueue: Array<IReactionEmojiProps>;
 }
 
+
+/**
+ * Implementation of a button for reactions.
+ */
+class ReactionsButtonImpl extends AbstractButton<AbstractButtonProps> {
```
Comment (robertpin) [inline react/features/reactions/components/web/ReactionsMenuButton.tsx:76]: I'm not a fan of this naming. Please leave it as `ReactionsButton` and extract the other component (`ReactionsMenuButton`) to another file

**Review by robertpin — CHANGES_REQUESTED** [review]

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx:43]: I don't see why it is weird to pass Objects as children. Could you please explain?

Correct me if this is wrong but what I've been told (or read) the `children` prop is like any other prop, isn't it? I've seen it widely used for all kind of types (in our project and in another projects as well).

The reason for the change is that I need to put the ReactionsMenu as a child of the ContextMenu that we are building on one side and on the other to not over-complicate the Toolbox component which have been already too big and complicated. Also IMHO I don't think that we need to keep the OverflowMenuButton so generic. Now we pass to it what needs to be rendered and it takes care how to render it.

File: react/features/gifs/components/web/GifsMenu.tsx (hunk 17 above)
Comment (hristoterezov) [reply react/features/gifs/components/web/GifsMenu.tsx:88]: As you can see from the type `parent` is not a component at all. `parent` is an enum that indicates where the component will be rendered - in drawer, in the overflow menu (not a drawer) or as part of a button.

In my initial implementation we had 2 props inDrawer and overflowMenu in order to establish where the component is rendered. If you check the discussion with @saghul from the previous comments I realized that this is not great and changed that to this enum.

The enum implementation is simpler and more readable.

File: react/features/reactions/components/web/ReactionsMenuButton.tsx (hunk 18 above)
Comment (hristoterezov) [reply react/features/reactions/components/web/ReactionsMenuButton.tsx:76]: I'm not a fan either. Now the problem is that we already have a file and component called `ReactionsButton`!

Also `ReactionsButtonImpl` will ever be needed only here. 

I was thinking about a better solution but failed to find one. If you have better idea please let me know!

File: react/features/reactions/components/web/ReactionsMenuButton.tsx [hunk 19]
```diff
@@ -1,16 +1,20 @@
...
 import { connect, useSelector } from 'react-redux';
 
 import { IReduxState } from '../../../app/types';
 import { isMobileBrowser } from '../../../base/environment/utils';
 import { translate } from '../../../base/i18n/functions';
-import { IconArrowUp } from '../../../base/icons/svg';
-import ToolboxButtonWithIconPopup from '../../../base/toolbox/components/web/ToolboxButtonWithIconPopup';
+import { IconArrowUp, IconFaceSmile } from '../../../base/icons/svg';
+// eslint-disable-next-line lines-around-comment
+// @ts-ignore
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/reactions/components/web/ReactionsMenuButton.tsx]: It turns out it is not! Good catch! It is removed now!

File: react/features/toolbox/components/web/OverflowMenuButton.tsx (hunk 12 above)
Comment (hristoterezov) [reply react/features/toolbox/components/web/OverflowMenuButton.tsx]: Now I'm passing the object array with the information about the buttons into normal prop (not children).

**Review by robertpin — APPROVED** [review]
Approved, but I have to say the reactions feature feels overly complicated now
